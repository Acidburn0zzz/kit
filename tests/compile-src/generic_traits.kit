trait GenericPrinter[T] {
    public function print(value: T): Void;
}

implement GenericPrinter[Float] for Int {
    public function print(value: Float) {
        printf("%.2f %.2f\n", this as Float, value);
    }
}

implement GenericPrinter[CString] for CString {
    public function print(value: CString) {
        printf("%s %s\n", this, value);
    }
}

function main() {
    var a: Int = 1;
    var aBox: Box[GenericPrinter[Float]] = a;
    aBox.print(2);

    var b = "hello";
    var bBox: Box[GenericPrinter[CString]] = b;
    bBox.print("world");

    var c = "hi";
    var cBox: Box[GenericPrinter] = c;
    cBox.print("there");
}
