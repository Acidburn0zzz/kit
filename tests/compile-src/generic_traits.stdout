1.00 2.00
hello world
hi there
